globus_cli/parsing/endpoint_plus_path.py

```diff
@@ -0,0 +1,62 @@
+import click
+
+
+class EndpointPlusPath(click.ParamType):
+    """
+    Custom type for "<endpoint_id>:<path>"
+    Supports path being required and path being optional.
+
+    Always produces a Tuple.
+    """
+    name = 'endpoint plus path'
+
+    def __init__(self, *args, **kwargs):
+        # path requirement defaults to True, but can be tweaked by a kwarg
+        self.path_required = kwargs.pop('path_required', True)
+
+        super(EndpointPlusPath, self).__init__(*args, **kwargs)
+
+    def get_metavar(self, param):
+        """
+        Default metavar for this instance of the type.
+        """
+        return self.metavar
+
+    @property
+    def metavar(self):
+        """
+        Metavar as a property, to satisfy the slight brokenness of
+        click.Argument
+        Tracked against Click as an issue:
+        https://github.com/pallets/click/issues/674
+        """
+        if self.path_required:
+            return "ENDPOINT_ID:PATH"
+        else:
+            return "ENDPOINT_ID[:PATH]"
+
+    def convert(self, value, param, ctx):
+        """
+        ParamType.convert() is the actual processing method that takes a
+        provided parameter and parses it.
+        """
+        # passthrough conditions: None or already processed
+        if value is None or isinstance(value, tuple):
+            return value
+
+        # split the value on the first colon, leave the rest intact
+        splitval = value.split(':', 1)
+        # first element is the endpoint_id
+        endpoint_id = splitval[0]
+
+        try:
+            path = splitval[1]
+        except IndexError:
+            path = None
+            if self.path_required:
+                self.fail(('This command requires a path in addition to an '
+                           'endpoint ID. You must use the '
+                           '"<endpoint_id>:<path>" notation.'),
+                          param=param)
```

Comment:
By only checking `self.path_required` on an `IndexError` exception, you allow something like `globus transfer async-delete ddb59aef-6d04-11e5-ba46-22000b92c6ec:` to sneak through, which could definitely happen if someone combines two input variables before calling the CLI.

Maybe something like this would be nice:
```
        try:
            path = splitval[1]
        except IndexError:
            path = None
        if self.path_required and not path:
            self.fail('This command requires a path in addition to an '
                      'endpoint ID. You must use the '
                      '"<endpoint_id>:<path>" notation.',
                      param=param)
```